<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Jam Base Methods for bioinformatics and data analysis • jamba</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Jam Base Methods for bioinformatics and data analysis">
<meta name="description" content="Efficient methods for analysis, mixedSort, lists, colors.">
<meta property="og:description" content="Efficient methods for analysis, mixedSort, lists, colors.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">jamba</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.107.900</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/jamba-vignette.html">Jamba Overview</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jmw86069/jamba/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="jamba">jamba<a class="anchor" aria-label="anchor" href="#jamba"></a>
</h1></div>
<p>The goal of jamba is to provide useful custom functions for R data analysis and visualization.</p>
<div class="section level2">
<h2 id="package-reference">Package Reference<a class="anchor" aria-label="anchor" href="#package-reference"></a>
</h2>
<p>A full online function reference is available via the pkgdown documentation:</p>
<p><a href="https://jmw86069.github.io/jamba/">Full jamba command reference</a></p>
<p>Functions are categorized, some examples are listed below:</p>
</div>
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>The R functions in <code>jamba</code> have been built up, used, tested, revised over several years. They are immediately useful for day-to-day work, and efficient and robust enough for production pipelines.</p>
<p>Many were inspired by discussion from Stackoverflow, R-help, or Bioconductor, with citations thanking principal author(s). Many thanks to the original authors! The R community is built upon the collective greatness of its contributors!</p>
<p>Most of the functions are designed around workflows for Bioinformatics analyses, where functions need to be efficient when operating over 10,000 to 100,000 elements. (They work quite well with millions as well.) Usually the speed gains are obvious with about 100 elements, then scale linearly (or worse) as the number increases. I and others use these functions <em>all the time</em>.</p>
<p>One example function <code><a href="reference/writeOpenxlsx.html">writeOpenxlsx()</a></code> is a simple wrapper around very useful <code><a href="https://rdrr.io/pkg/openxlsx/man/write.xlsx.html" class="external-link">openxlsx::write.xlsx()</a></code>, which also applies column formatting for column types: P-values, fold changes, log2 fold changes, numeric, and integer values. Columns use conditional Excel formatting to apply color-shading to cells for each type.</p>
<p>Similarly, <code><a href="reference/readOpenxlsx.html">readOpenxlsx()</a></code> is a wrapper function to <code><a href="https://rdrr.io/pkg/openxlsx/man/read.xlsx.html" class="external-link">openxlsx::read.xlsx()</a></code> which reads each worksheet and returns a <code>list</code> of <code>data.frame</code> objects. It can detect multi-row column headers, for which it returns combined column names. It also applies equivalent of <code>check.names=FALSE</code> so column names are returned without change.</p>
<p>Small and large efficiencies are used wherever possible. The <code><a href="reference/mixedSort.html">mixedSort()</a></code> functions are based upon <code><a href="https://rdrr.io/pkg/gtools/man/mixedsort.html" class="external-link">gtools::mixedsort()</a></code>, with additional optimizations for speed and custom needs. It sorts chromosome names, gene names, micro-RNA names, etc.</p>
</div>
<div class="section level2">
<h2 id="example-r-functions">Example R functions<a class="anchor" aria-label="anchor" href="#example-r-functions"></a>
</h2>
<div class="section level3">
<h3 id="efficient-alphanumeric-sort">Efficient alphanumeric sort<a class="anchor" aria-label="anchor" href="#efficient-alphanumeric-sort"></a>
</h3>
<ul>
<li>
<code><a href="reference/mixedSort.html">mixedSort()</a></code> - highly efficient alphanumeric sort, for example chr1, chr2, chr3, chr10, etc.</li>
<li>
<code><a href="reference/mixedSortDF.html">mixedSortDF()</a></code> - as above, applied to columns in a <code>data.frame</code> (or <code>matrix</code>, <code>tibble</code>, <code>DataFrame</code>, etc.)</li>
<li>
<code><a href="reference/mixedSorts.html">mixedSorts()</a></code> - as above, applied to a list of vectors with no speed loss.</li>
</ul>
<p>Example:</p>
<div class="kable-table">
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="left">miRNA</th>
<th align="right">sort_rank</th>
<th align="right">mixedSort_rank</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">2</td>
<td align="left">ABCA2</td>
<td align="right">2</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="left">ABCA12</td>
<td align="right">1</td>
<td align="right">2</td>
</tr>
<tr class="odd">
<td align="left">3</td>
<td align="left">miR-1</td>
<td align="right">3</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">6</td>
<td align="left">miR-1a</td>
<td align="right">6</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">7</td>
<td align="left">miR-1b</td>
<td align="right">7</td>
<td align="right">5</td>
</tr>
<tr class="even">
<td align="left">8</td>
<td align="left">miR-2</td>
<td align="right">8</td>
<td align="right">6</td>
</tr>
<tr class="odd">
<td align="left">4</td>
<td align="left">miR-12</td>
<td align="right">4</td>
<td align="right">7</td>
</tr>
<tr class="even">
<td align="left">9</td>
<td align="left">miR-22</td>
<td align="right">9</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="left">5</td>
<td align="left">miR-122</td>
<td align="right">5</td>
<td align="right">9</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level3">
<h3 id="base-r-plotting">Base R plotting<a class="anchor" aria-label="anchor" href="#base-r-plotting"></a>
</h3>
<p>These functions help with base R plots, in all those little cases when the amazing <code>ggplot2</code> package is not a smooth fit.</p>
<ul>
<li>
<code><a href="reference/nullPlot.html">nullPlot()</a></code> - convenient “blank” base R plot, optionally displays margins</li>
<li>
<code><a href="reference/plotSmoothScatter.html">plotSmoothScatter()</a></code> - smooth scatter <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> for point density, enhanced over <code><a href="https://rdrr.io/r/graphics/smoothScatter.html" class="external-link">smoothScatter()</a></code> <img src="reference/figures/README-plotSmoothScatter-1.png" alt="Example of smooth scatterplots using the default R function, and enhanced function in the jamba package.">
</li>
<li>
<code><a href="reference/plotPolygonDensity.html">plotPolygonDensity()</a></code> - fast density/histogram plot for vector or matrix <img src="reference/figures/README-plotPolygonDensity-1.png" alt="Example polygon density plot.">
</li>
<li>
<code><a href="reference/imageDefault.html">imageDefault()</a></code> - enhanced <code><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image()</a></code> that enables raster output with consistent pixel aspect ratio.</li>
<li>
<code><a href="reference/imageByColors.html">imageByColors()</a></code> - wrapper to <code><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image()</a></code> for a matrix or data.frame of colors, with optional labels <img src="reference/figures/README-imageByColors-1.png" alt="Example color matrix as plotted using the image by colors function()">
</li>
<li>
<code><a href="reference/minorLogTicksAxis.html">minorLogTicksAxis()</a></code> - log-transformed axis labels, flexible log base, and option for properly adjusted <code>log2(1 + x)</code> format.</li>
<li>
<code><a href="reference/sqrtAxis.html">sqrtAxis()</a></code> - draw a square-root transformed axis, with proper labels.</li>
<li>
<code><a href="reference/drawLabels.html">drawLabels()</a></code> - draw square colorized text labels</li>
<li>
<code><a href="reference/shadowText.html">shadowText()</a></code> - replacement for <code><a href="https://rdrr.io/r/graphics/text.html" class="external-link">text()</a></code> that draws shadows or outlines. <img src="reference/figures/README-labels-1.png" alt="Example showing text labels, and shadow text labels on a colored background.">
</li>
<li>
<code><a href="reference/groupedAxis.html">groupedAxis()</a></code> - grouped axis labels to show regions/ranges</li>
<li>
<code><a href="reference/decideMfrow.html">decideMfrow()</a></code> - determine appropriate value for <code>par("mfrow")</code> for multipanel output in base R plotting.</li>
<li>
<code><a href="reference/getPlotAspect.html">getPlotAspect()</a></code> - determine visible plot aspect ratio.</li>
</ul>
</div>
<div class="section level3">
<h3 id="excel-export">Excel export<a class="anchor" aria-label="anchor" href="#excel-export"></a>
</h3>
<p>Every Bioinformatician/statistician needs to write data to Excel, the <code><a href="reference/writeOpenxlsx.html">writeOpenxlsx()</a></code> function is consistent and makes it look pretty. You can save numerous worksheets in a single Excel file, without having to go back and custom-format everything.</p>
<ul>
<li>
<code><a href="reference/writeOpenxlsx.html">writeOpenxlsx()</a></code> - flexible Excel exporter, with categorical and conditional colors.</li>
<li>
<code><a href="reference/applyXlsxCategoricalFormat.html">applyXlsxCategoricalFormat()</a></code> - apply categorical colors to Excel</li>
<li>
<code><a href="reference/applyXlsxConditionalFormat.html">applyXlsxConditionalFormat()</a></code> - apply conditional colors to Excel</li>
</ul>
</div>
<div class="section level3">
<h3 id="color">Color<a class="anchor" aria-label="anchor" href="#color"></a>
</h3>
<p>Almost everything uses color somewhere, especially on R console, and in every R plot.</p>
<ul>
<li>
<code><a href="reference/getColorRamp.html">getColorRamp()</a></code> - flexible to create or retrieve color gradients</li>
<li>
<code><a href="reference/warpRamp.html">warpRamp()</a></code> - “bend” a color gradient to enhance the visual range</li>
<li>
<code><a href="reference/color2gradient.html">color2gradient()</a></code> - convert a color to gradient of n colors; or do the same for a vector</li>
<li>
<code><a href="reference/makeColorDarker.html">makeColorDarker()</a></code> - adjust darkness and saturation</li>
<li>
<code><a href="reference/showColors.html">showColors()</a></code> - display a vector or list of colors</li>
<li>
<code><a href="reference/fixYellow.html">fixYellow()</a></code> - adjust the weird green-yellow, by personal preference</li>
<li>
<code><a href="reference/printDebug.html">printDebug()</a></code> - pretty colorized text output using <code>crayon</code> package.</li>
<li>
<code><a href="reference/rainbow2.html">rainbow2()</a></code> - rainbow categorical colors with enhanced visual contrast.</li>
</ul>
<p><img src="reference/figures/README-colorshow-1.png" alt="Image showing a series of color palettes, adjusting contrast with lens, and expanding palettes with color2gradient()"></p>
</div>
<div class="section level3">
<h3 id="list">List<a class="anchor" aria-label="anchor" href="#list"></a>
</h3>
<p>Cool methods to operate on super-long lists in one call, to avoid looping through the list either with <code>for()</code> loops, <code><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply()</a></code> or <code>map()</code> functions.</p>
<ul>
<li>
<code><a href="reference/cPaste.html">cPaste()</a></code> - highly efficient <code><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste()</a></code> over a large list of vectors</li>
<li>
<code><a href="reference/cPasteS.html">cPasteS()</a></code> - as above but using <code><a href="reference/mixedSort.html">mixedSort()</a></code> before <code><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste()</a></code>.</li>
<li>
<code><a href="reference/cPasteU.html">cPasteU()</a></code> - as above but using <code><a href="reference/uniques.html">uniques()</a></code> before <code><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste()</a></code>.</li>
<li>
<code><a href="reference/cPasteSU.html">cPasteSU()</a></code> - as above but using <code><a href="reference/mixedSort.html">mixedSort()</a></code> and <code><a href="reference/uniques.html">uniques()</a></code> before <code><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste()</a></code>.</li>
<li>
<code><a href="reference/uniques.html">uniques()</a></code> - efficient <code><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique()</a></code> over a list of vectors</li>
<li>
<code><a href="reference/sclass.html">sclass()</a></code> - runs <code><a href="https://rdrr.io/r/base/class.html" class="external-link">class()</a></code> on a list</li>
<li>
<code><a href="reference/sdim.html">sdim()</a></code>, <code><a href="reference/sdim.html">ssdim()</a></code> - dimensions of list objects, or nested list of lists</li>
<li>
<code><a href="reference/rbindList.html">rbindList()</a></code> - efficient <code>do.call(rbind, ...)</code> to bind rows into a matrix or data.frame, useful when following <code><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit()</a></code>.</li>
<li>
<code><a href="reference/mergeAllXY.html">mergeAllXY()</a></code> - merge a list of <code>data.frame</code> objects</li>
<li>
<code><a href="reference/rmNULL.html">rmNULL()</a></code> - remove NULL or empty elements from a list, with optional replacement</li>
</ul>
</div>
<div class="section level3">
<h3 id="names">Names<a class="anchor" aria-label="anchor" href="#names"></a>
</h3>
<p>We use R names as an additional method to make sure everything is kept in the proper order. Many R functions return results using input names, so it helps to have a really solid naming strategy. For the R functions that remove names – I highly recommend adding them back yourself!</p>
<ul>
<li>
<code><a href="reference/makeNames.html">makeNames()</a></code> - make unique names, using flexible logic</li>
<li>
<code><a href="reference/nameVector.html">nameVector()</a></code> - add names to a vector, using its own value, or supplied names</li>
<li>
<code><a href="reference/nameVectorN.html">nameVectorN()</a></code> - make named vector using the names of a vector (useful inside <code><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply()</a></code>) or any function that returns data using names of the input vector.</li>
</ul>
</div>
<div class="section level3">
<h3 id="dataframematrixtibble">data.frame/matrix/tibble<a class="anchor" aria-label="anchor" href="#dataframematrixtibble"></a>
</h3>
<ul>
<li>
<code><a href="reference/pasteByRow.html">pasteByRow()</a></code> - fast, flexible row-paste with delimiters, optionally remove blanks</li>
<li>
<code><a href="reference/pasteByRowOrdered.html">pasteByRowOrdered()</a></code> - as above but returns ordered factor, using existing factor orders from each column when present</li>
<li>
<code><a href="reference/rowGroupMeans.html">rowGroupMeans()</a></code>, <code><a href="reference/rowRmMadOutliers.html">rowRmMadOutliers()</a></code> - efficient grouped row functions</li>
<li>
<code><a href="reference/mergeAllXY.html">mergeAllXY()</a></code> - merge a list of <code>data.frame</code> into one</li>
<li>
<code><a href="reference/renameColumn.html">renameColumn()</a></code> - rename columns <code>from</code> and <code>to</code>.</li>
<li>
<code><a href="reference/kable_coloring.html">kable_coloring()</a></code> - flexible colorized <code>data.frame</code> output in Rmarkdown.</li>
</ul>
</div>
<div class="section level3">
<h3 id="string--grep">String / grep<a class="anchor" aria-label="anchor" href="#string--grep"></a>
</h3>
<ul>
<li>
<code><a href="reference/gsubOrdered.html">gsubOrdered()</a></code> - gsub that returns ordered factor, maintians the previous factor order</li>
<li>
<code><a href="reference/grepls.html">grepls()</a></code> - grep the environment (including attached packages) for object names</li>
<li>
<code><a href="reference/vgrep.html">vgrep()</a></code>, <code><a href="reference/vigrep.html">vigrep()</a></code> - value-grep shortcut</li>
<li>
<code>unvgrep()</code>, <code><a href="reference/unvigrep.html">unvigrep()</a></code> - un-grep – remove matched results from the output.</li>
<li>
<code><a href="reference/provigrep.html">provigrep()</a></code> - progressive grep, searches each pattern in order, returning results in that order</li>
<li>
<code><a href="reference/igrepHas.html">igrepHas()</a></code> - rapid case-insensitive grep presence/absense test</li>
<li>
<code><a href="reference/ucfirst.html">ucfirst()</a></code> - upper-case the first letter of each word.</li>
<li>
<code><a href="reference/padString.html">padString()</a></code>, <code><a href="reference/padInteger.html">padInteger()</a></code> - produce strings from numeric values with consistent leading zeros.</li>
</ul>
</div>
<div class="section level3">
<h3 id="numeric">Numeric<a class="anchor" aria-label="anchor" href="#numeric"></a>
</h3>
<ul>
<li>
<code><a href="reference/noiseFloor.html">noiseFloor()</a></code> - apply noise floor (and ceiling) with flexible replacement values</li>
<li>
<code><a href="reference/warpAroundZero.html">warpAroundZero()</a></code> - warp a numeric vector symmetrically around zero</li>
<li>
<code><a href="reference/rowGroupMeans.html">rowGroupMeans()</a></code>, <code><a href="reference/rowRmMadOutliers.html">rowRmMadOutliers()</a></code> - efficient grouped row functions</li>
<li>
<code><a href="reference/deg2rad.html">deg2rad()</a></code>, <code><a href="reference/rad2deg.html">rad2deg()</a></code> - convert degrees to radians</li>
<li>
<code><a href="reference/rmNA.html">rmNA()</a></code> - remove NA values, with optional replacement</li>
<li>
<code><a href="reference/rmInfinite.html">rmInfinite()</a></code> - remove infinite values, with optional replacement.</li>
<li>
<code><a href="reference/formatInt.html">formatInt()</a></code> - convenient <code><a href="https://rdrr.io/r/base/format.html" class="external-link">format()</a></code> for integer output, with comma-delimiter by default</li>
</ul>
</div>
<div class="section level3">
<h3 id="practical--helpful">Practical / helpful<a class="anchor" aria-label="anchor" href="#practical--helpful"></a>
</h3>
<ul>
<li>
<code><a href="reference/jargs.html">jargs()</a></code> - pretty function arguments, optional pattern search argument name</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/jargs.html">jargs</a></span><span class="op">(</span><span class="va">plotSmoothScatter</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 x = ,</span></span>
<span><span class="co">#&gt;                 y = NULL,</span></span>
<span><span class="co">#&gt;              bwpi = 50,</span></span>
<span><span class="co">#&gt;             binpi = 50,</span></span>
<span><span class="co">#&gt;        bandwidthN = NULL,</span></span>
<span><span class="co">#&gt;              nbin = NULL,</span></span>
<span><span class="co">#&gt;            expand = c(0.04, 0.04),</span></span>
<span><span class="co">#&gt;       transFactor = 0.25,</span></span>
<span><span class="co">#&gt;    transformation = function( x ) x^transFactor,</span></span>
<span><span class="co">#&gt;              xlim = NULL,</span></span>
<span><span class="co">#&gt;              ylim = NULL,</span></span>
<span><span class="co">#&gt;              xlab = NULL,</span></span>
<span><span class="co">#&gt;              ylab = NULL,</span></span>
<span><span class="co">#&gt;          nrpoints = 0,</span></span>
<span><span class="co">#&gt;           colramp = c("white", "lightblue", "blue", "orange", "orangered2"),</span></span>
<span><span class="co">#&gt;               col = "black",</span></span>
<span><span class="co">#&gt;            doTest = FALSE,</span></span>
<span><span class="co">#&gt;    fillBackground = TRUE,</span></span>
<span><span class="co">#&gt;          naAction = c("remove", "floor0", "floor1"),</span></span>
<span><span class="co">#&gt;              xaxt = "s",</span></span>
<span><span class="co">#&gt;              yaxt = "s",</span></span>
<span><span class="co">#&gt;               add = FALSE,</span></span>
<span><span class="co">#&gt;               asp = NULL,</span></span>
<span><span class="co">#&gt; applyRangeCeiling = TRUE,</span></span>
<span><span class="co">#&gt;         useRaster = TRUE,</span></span>
<span><span class="co">#&gt;           verbose = FALSE,</span></span>
<span><span class="co">#&gt;               ... =</span></span></code></pre></div>
<ul>
<li>
<code><a href="reference/sdim.html">sdim()</a></code>, <code><a href="reference/sdim.html">ssdim()</a></code> - dimensions of list objects, or nested list of lists</li>
<li>
<code><a href="reference/sdim.html">sdima()</a></code> - runs <code><a href="reference/sdim.html">sdim()</a></code> on the attributes of an object.</li>
<li>
<code><a href="reference/isTRUEV.html">isTRUEV()</a></code>, <code><a href="reference/isFALSEV.html">isFALSEV()</a></code> - vectorized test for TRUE or FALSE values, since <code><a href="https://rdrr.io/r/base/Logic.html" class="external-link">isTRUE()</a></code> only operates on single values, and does not allow <code>NA</code>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="r-console">R console<a class="anchor" aria-label="anchor" href="#r-console"></a>
</h3>
<ul>
<li>
<code><a href="reference/printDebug.html">printDebug()</a></code> - pretty colorized text output using <code>crayon</code> package.</li>
<li>
<code><a href="reference/setPrompt.html">setPrompt()</a></code> - pretty colorized R console prompt with project name and R version</li>
<li>
<code><a href="reference/newestFile.html">newestFile()</a></code> - most recently modified file from a vector of files</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="other-related-jam-packages">Other related Jam packages<a class="anchor" aria-label="anchor" href="#other-related-jam-packages"></a>
</h2>
<ul>
<li>
<code>jamma</code> – MA-plots (also known as “mean-variance”, “Bland-Altman”, or “mean-difference” plots), relies upon <code><a href="reference/plotSmoothScatter.html">jamba::plotSmoothScatter()</a></code>; <code>centerGeneData()</code> to apply flexible row-centering with optional groups and control samples; <code>jammanorm()</code> - normalize data based upon MA-plot output</li>
<li>
<code>colorjam</code> – <code><a href="https://jmw86069.github.io/colorjam/reference/rainbowJam.html" class="external-link">colorjam::rainbowJam()</a></code> for scalable categorical colors using alternating luminance and chroma values.</li>
<li>
<code>genejam</code> – fast, consistent conversion of gene symbols to the most current gene nomenclature</li>
<li>
<code>splicejam</code> – Sashimi plots for RNA-seq data</li>
<li>
<code>multienrichjam</code> – multiple gene set enrichment analysis and visualization</li>
<li>
<code>platjam</code> – platform technology functions, importers for NanoString</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/jmw86069/jamba/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/jmw86069/jamba/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing jamba</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>James M. Ward <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by James M. Ward.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
