<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Jam Base Methods for bioinformatics and data analysis • jamba</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Jam Base Methods for bioinformatics and data analysis">
<meta name="description" content="Efficient methods for analysis, mixedSort, lists, colors.">
<meta property="og:description" content="Efficient methods for analysis, mixedSort, lists, colors.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">jamba</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/jamba-vignette.html">Jamba Overview</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jmw86069/jamba/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="jamba">jamba<a class="anchor" aria-label="anchor" href="#jamba"></a>
</h1></div>
<p>The goal of jamba is to provide useful custom functions for R data analysis and visualization. jamba version <strong>1.0.2</strong></p>
<div class="section level2">
<h2 id="package-reference">Package Reference<a class="anchor" aria-label="anchor" href="#package-reference"></a>
</h2>
<p>A full online function reference is available via the pkgdown documentation:</p>
<p><a href="https://jmw86069.github.io/jamba/">Full jamba command reference</a></p>
<p>Functions are categorized, some examples are listed below:</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Production will soon be available from CRAN:</p>
<p><code>install.packages("jamba")</code></p>
<p>The development version can be installed:</p>
<p><code>remotes::install_github("jmw86069/jamba")</code></p>
<div class="section level3">
<h3 id="additional-r-packages-in-suggests">Additional R Packages in “Suggests”<a class="anchor" aria-label="anchor" href="#additional-r-packages-in-suggests"></a>
</h3>
<ul>
<li>
<code>crayon</code> - install with <code>install.packages("crayon")</code> for glorious colored console output. Color makes it better.</li>
<li>
<code>farver</code> - install with <code>install.packages("farver")</code> for more efficient color manipulations, and HSL color coneversions.</li>
</ul>
</div>
<div class="section level3">
<h3 id="additional-r-packages-in-enhances">Additional R Packages in “Enhances”<a class="anchor" aria-label="anchor" href="#additional-r-packages-in-enhances"></a>
</h3>
<p>Bioconductor packages are invaluable for bioinformatics work, but can be a bit “heavy” to install if not absolutely necessary. Therefore, Bioconductor packages are in “Enhances” so they require someone to make the choice to install them.</p>
<ul>
<li>
<code>S4Vectors</code> - install with <code>BiocManager::install("S4vectors")</code> to improve speed of <code><a href="reference/cPaste.html">cPaste()</a></code> functions.</li>
<li>
<code>openxlsx</code> - install with <code>install.packages("openxlsx")</code> to support Excel <code>xlsx</code> file import, and stylized export.</li>
<li>
<code>kableExtra</code> - install with <code>install.packages("kableExtra")</code> to enable colorized kable HTML tables in RMarkdown documents.</li>
<li>
<code>ComplexHeatmap</code> - install with <code>BiocManager::install("ComplexHeatmap")</code> to use with <code><a href="reference/heatmap_row_order.html">heatmap_row_order()</a></code>, <code><a href="reference/cell_fun_label.html">cell_fun_label()</a></code> for custom labels.</li>
<li>
<code>matrixStats</code> - install with <code>install.packages("matrixStats")</code> for efficient <code>numeric</code> stats calculations, or <code>sparseMatrixStats</code> for use with Matrix sparse matrices as used with Seurat and SingleCellExperiment data.</li>
<li>
<code>ggridges</code> - install with <code>install.packages("ggridges")</code> for convenient ridge density plots using <code><a href="reference/plotRidges.html">plotRidges()</a></code>.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>The R functions in <code>jamba</code> have been built up, used, tested, revised over several years. They are immediately useful for day-to-day work, and efficient and robust enough for production pipelines.</p>
<p>Many were inspired by discussion from Stackoverflow, R-help, or Bioconductor, with citations thanking principal author(s). Many thanks to the original authors! The R community is built upon the collective greatness of its contributors!</p>
<p>Most of the functions are designed around workflows for Bioinformatics analyses, where functions need to be efficient when operating over 10,000 to 100,000 elements. (They work quite well with millions as well.) Usually the speed gains are obvious with about 100 elements, then scale linearly (or worse) as the number increases. I and others use these functions <em>all the time</em>.</p>
<p>One example function <code><a href="reference/writeOpenxlsx.html">writeOpenxlsx()</a></code> is a simple wrapper around very useful <code><a href="https://rdrr.io/pkg/openxlsx/man/write.xlsx.html" class="external-link">openxlsx::write.xlsx()</a></code>, which also applies column formatting for column types: P-values, fold changes, log2 fold changes, numeric, and integer values. Columns use conditional Excel formatting to apply color-shading to cells for each type.</p>
<p>Similarly, <code><a href="reference/readOpenxlsx.html">readOpenxlsx()</a></code> is a wrapper function to <code><a href="https://rdrr.io/pkg/openxlsx/man/read.xlsx.html" class="external-link">openxlsx::read.xlsx()</a></code> which reads each worksheet and returns a <code>list</code> of <code>data.frame</code> objects. It can detect multi-row column headers, for which it returns combined column names. It also applies equivalent of <code>check.names=FALSE</code> so column names are returned without change.</p>
<p>Small and large efficiencies are used wherever possible. The <code><a href="reference/mixedSort.html">mixedSort()</a></code> functions are based upon <code><a href="https://rdrr.io/pkg/gtools/man/mixedsort.html" class="external-link">gtools::mixedsort()</a></code>, with additional optimizations for speed and custom needs. It sorts chromosome names, gene names, micro-RNA names, etc.</p>
</div>
<div class="section level2">
<h2 id="alphanumeric-sort">Alphanumeric sort<a class="anchor" aria-label="anchor" href="#alphanumeric-sort"></a>
</h2>
<ul>
<li>
<code><a href="reference/mixedSort.html">mixedSort()</a></code> - highly efficient alphanumeric sort, for example chr1, chr2, chr3, chr10, etc.</li>
<li>
<code><a href="reference/mixedSortDF.html">mixedSortDF()</a></code> - as above, applied to columns in a <code>data.frame</code> (or <code>matrix</code>, <code>tibble</code>, <code>DataFrame</code>, etc.)</li>
<li>
<code><a href="reference/mixedSorts.html">mixedSorts()</a></code> - as above, applied to a list of vectors with no speed loss.</li>
</ul>
<p>Example:</p>
<div class="kable-table">
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="left">miRNA</th>
<th align="right">sort_rank</th>
<th align="right">mixedSort_rank</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">2</td>
<td align="left">ABCA2</td>
<td align="right">2</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="left">ABCA12</td>
<td align="right">1</td>
<td align="right">2</td>
</tr>
<tr class="odd">
<td align="left">3</td>
<td align="left">miR-1</td>
<td align="right">3</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">6</td>
<td align="left">miR-1a</td>
<td align="right">6</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">7</td>
<td align="left">miR-1b</td>
<td align="right">7</td>
<td align="right">5</td>
</tr>
<tr class="even">
<td align="left">8</td>
<td align="left">miR-2</td>
<td align="right">8</td>
<td align="right">6</td>
</tr>
<tr class="odd">
<td align="left">4</td>
<td align="left">miR-12</td>
<td align="right">4</td>
<td align="right">7</td>
</tr>
<tr class="even">
<td align="left">9</td>
<td align="left">miR-22</td>
<td align="right">9</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="left">5</td>
<td align="left">miR-122</td>
<td align="right">5</td>
<td align="right">9</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="base-r-plotting">Base R plotting<a class="anchor" aria-label="anchor" href="#base-r-plotting"></a>
</h2>
<p>These functions help with base R plots, in all those little cases when the amazing <code>ggplot2</code> package is not a smooth fit.</p>
<ul>
<li>
<code><a href="reference/nullPlot.html">nullPlot()</a></code> - convenient “blank” base R plot, optionally displays margins</li>
<li>
<code><a href="reference/plotSmoothScatter.html">plotSmoothScatter()</a></code> - smooth scatter <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> for point density, enhanced over <code><a href="https://rdrr.io/r/graphics/smoothScatter.html" class="external-link">smoothScatter()</a></code> <img src="reference/figures/README-plotSmoothScatter-1.png" alt="Example of smooth scatterplots using the default R function, and enhanced function in the jamba package.">
</li>
<li>
<code><a href="reference/plotPolygonDensity.html">plotPolygonDensity()</a></code> - fast density/histogram plot for vector or matrix <img src="reference/figures/README-plotPolygonDensity-1.png" alt="Example polygon density plot.">
</li>
<li>
<code><a href="reference/imageDefault.html">imageDefault()</a></code> - enhanced <code><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image()</a></code> that enables raster output with consistent pixel aspect ratio.</li>
<li>
<code><a href="reference/imageByColors.html">imageByColors()</a></code> - wrapper to <code><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image()</a></code> for a matrix or data.frame of colors, with optional labels <img src="reference/figures/README-imageByColors-1.png" alt="Example color matrix as plotted using the image by colors function()">
</li>
<li>
<code><a href="reference/minorLogTicksAxis.html">minorLogTicksAxis()</a></code>, <code><a href="reference/minorLogTicksAxis.html">logFoldAxis()</a></code>, <code><a href="reference/minorLogTicksAxis.html">pvalueAxis()</a></code> - log axis tick marks and labels, compatible with <code>offset</code> for example <code>log(offset + x)</code>.</li>
<li>
<code><a href="reference/sqrtAxis.html">sqrtAxis()</a></code> - draw a square-root transformed axis, with proper labels.</li>
<li>
<code><a href="reference/drawLabels.html">drawLabels()</a></code> - draw square colorized text labels</li>
<li>
<code><a href="reference/shadowText.html">shadowText()</a></code> - replacement for <code><a href="https://rdrr.io/r/graphics/text.html" class="external-link">text()</a></code> that draws shadows or outlines. <img src="reference/figures/README-labels-1.png" alt="Example showing text labels, and shadow text labels on a colored background.">
</li>
<li>
<code><a href="reference/groupedAxis.html">groupedAxis()</a></code> - grouped axis labels to show regions/ranges</li>
<li>
<code><a href="reference/decideMfrow.html">decideMfrow()</a></code> - determine appropriate value for <code>par("mfrow")</code> for multipanel output in base R plotting.</li>
<li>
<code><a href="reference/getPlotAspect.html">getPlotAspect()</a></code> - determine visible plot aspect ratio.</li>
</ul>
</div>
<div class="section level2">
<h2 id="excel-export">Excel export<a class="anchor" aria-label="anchor" href="#excel-export"></a>
</h2>
<p>Every Bioinformatician/statistician needs to write data to Excel, the <code><a href="reference/writeOpenxlsx.html">writeOpenxlsx()</a></code> function is consistent and makes it look pretty. You can save numerous worksheets in a single Excel file, without having to go back and custom-format everything.</p>
<ul>
<li>
<code><a href="reference/writeOpenxlsx.html">writeOpenxlsx()</a></code> - flexible Excel exporter, with categorical and conditional colors.</li>
<li>
<code><a href="reference/applyXlsxCategoricalFormat.html">applyXlsxCategoricalFormat()</a></code> - apply categorical colors to Excel</li>
<li>
<code><a href="reference/applyXlsxConditionalFormat.html">applyXlsxConditionalFormat()</a></code> - apply conditional colors to Excel</li>
</ul>
</div>
<div class="section level2">
<h2 id="colors">Colors<a class="anchor" aria-label="anchor" href="#colors"></a>
</h2>
<p>Almost everything uses color somewhere, especially on R console, and in every R plot.</p>
<ul>
<li>
<code><a href="reference/getColorRamp.html">getColorRamp()</a></code> - retrieve or create color palettes</li>
<li>
<code><a href="reference/setTextContrastColor.html">setTextContrastColor()</a></code> - find contrasting font color for colored background</li>
<li>
<code><a href="reference/makeColorDarker.html">makeColorDarker()</a></code> - make a color darker (or lighter, or saturated)</li>
<li>
<code><a href="reference/color2gradient.html">color2gradient()</a></code> - split one color to a gradient of <code>n</code> colors</li>
<li>
<code><a href="reference/showColors.html">showColors()</a></code> - display a vector or <code>list</code> of colors</li>
<li>
<code><a href="reference/rainbow2.html">rainbow2()</a></code> - enhances <code><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">rainbow()</a></code> categorical colors for visual contrast.</li>
<li>
<code><a href="reference/warpRamp.html">warpRamp()</a></code> - “bend” a color gradient to enhance the visual range</li>
<li>
<code><a href="reference/fixYellow.html">fixYellow()</a></code> - opinionated reduction of yellow-green hue</li>
<li>
<code><a href="reference/printDebug.html">printDebug()</a></code> - colorized text output to console or RMarkdown</li>
<li>
<code><a href="reference/printDebug.html">printDebugHtml()</a></code> - colorized HTML output in RMarkdown or web pages</li>
<li>
<code><a href="reference/kable_coloring.html">kable_coloring()</a></code> - colored <code><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kableExtra::kable()</a></code> RMarkdown tables, if <code>kableExtra</code> package is installed.</li>
<li>
<code><a href="reference/col2alpha.html">col2alpha()</a></code>, <code><a href="reference/alpha2col.html">alpha2col()</a></code> - get or set alpha transparency</li>
<li>
<code><a href="reference/col2hcl.html">col2hcl()</a></code>, <code><a href="reference/col2hsl.html">col2hsl()</a></code>, <code><a href="reference/col2hsv.html">col2hsv()</a></code>, <code><a href="reference/hcl2col.html">hcl2col()</a></code>, <code><a href="reference/hsl2col.html">hsl2col()</a></code>, <code><a href="reference/hsv2col.html">hsv2col()</a></code>, <code><a href="reference/rgb2col.html">rgb2col()</a></code> - consistent color conversions.</li>
<li>
<code><a href="reference/color_dither.html">color_dither()</a></code> - split color into two to make color stripes</li>
</ul>
<p><img src="reference/figures/README-colorshow-1.png" alt="Image showing a series of color palettes, adjusting contrast with lens, and expanding palettes with color2gradient()"></p>
</div>
<div class="section level2">
<h2 id="list-functions">List Functions<a class="anchor" aria-label="anchor" href="#list-functions"></a>
</h2>
<p>Efficient methods to operate on lists in one call, to avoid looping through the list either with <code>for()</code> loops, <code><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply()</a></code> or <code>map()</code> functions. Driven by speed with 10k-100k rows, typical biological datasets.</p>
<p>Compared to convenient alternatives, <code><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply()</a></code> or tidyverse, typically order of magnitude faster. (Ymmv.) Notable exceptions: <code>data.table</code> and Bioconductor <code>S4Vectors</code>. Both are amazing, and are fairly heavy installations. <code>S4Vectors</code> is used when available.</p>
<ul>
<li>
<code><a href="reference/cPaste.html">cPaste()</a></code> - <code>paste(..., collapse)</code> a list of vectors</li>
<li>
<code><a href="reference/cPaste.html">cPasteS()</a></code> - <code><a href="reference/cPaste.html">cPaste()</a></code> with <code><a href="reference/mixedSort.html">mixedSort()</a></code>
</li>
<li>
<code><a href="reference/cPaste.html">cPasteU()</a></code> - <code><a href="reference/cPaste.html">cPaste()</a></code> with <code><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique()</a></code> (actually <code><a href="reference/uniques.html">uniques()</a></code>)</li>
<li>
<code><a href="reference/cPaste.html">cPasteSU()</a></code> - <code><a href="reference/cPaste.html">cPaste()</a></code> with <code><a href="reference/mixedSort.html">mixedSort()</a></code> and <code><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique()</a></code>
</li>
<li>
<code><a href="reference/uniques.html">uniques()</a></code> - <code><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique()</a></code> across a list of vectors</li>
<li>
<code><a href="reference/sclass.html">sclass()</a></code> - <code><a href="https://rdrr.io/r/base/class.html" class="external-link">class()</a></code> a list</li>
<li>
<code><a href="reference/sdim.html">sdim()</a></code> - <code><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim()</a></code> across a list, or S4 object, or non-list object</li>
<li>
<code><a href="reference/sdim.html">ssdim()</a></code> - <code><a href="reference/sdim.html">sdim()</a></code> across a list</li>
<li>
<code><a href="reference/sdim.html">sdima()</a></code> - <code><a href="reference/sdim.html">sdim()</a></code> for <code><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes()</a></code>
</li>
<li>
<code><a href="reference/rbindList.html">rbindList()</a></code> - <code>do.call(rbind, ...)</code> to bind rows into a <code>matrix</code> or <code>data.frame</code>, useful together with <code><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit()</a></code>.</li>
<li>
<code><a href="reference/mergeAllXY.html">mergeAllXY()</a></code> - <code>merge(..., all.x=TRUE, all.y=TRUE)</code> a list of <code>data.frame</code>
</li>
<li>
<code><a href="reference/rmNULL.html">rmNULL()</a></code> - remove NULL from a list, with optional replacement</li>
<li>
<code><a href="reference/rmNAs.html">rmNAs()</a></code> - <code><a href="reference/rmNA.html">rmNA()</a></code> across a list, with option replacement(s)</li>
<li>
<code><a href="reference/showColors.html">showColors()</a></code> - display colors</li>
<li>
<code><a href="reference/heads.html">heads()</a></code> - <code><a href="https://rdrr.io/r/utils/head.html" class="external-link">head()</a></code> across a list</li>
</ul>
</div>
<div class="section level2">
<h2 id="unique-names-with-versions">Unique names with versions<a class="anchor" aria-label="anchor" href="#unique-names-with-versions"></a>
</h2>
<p>R object names provide an additional method to confirm data are kept in the proper order. Duplicated names may be silently ignored, which motivated the easy approach to “make unique names”.</p>
<ul>
<li>
<code><a href="reference/makeNames.html">makeNames()</a></code> - make unique names, with flexible rules</li>
<li>
<code><a href="reference/nameVector.html">nameVector()</a></code> - add unique names using <code><a href="reference/makeNames.html">makeNames()</a></code>
</li>
<li>
<code><a href="reference/nameVectorN.html">nameVectorN()</a></code> - make vector of names, named with <code><a href="reference/makeNames.html">makeNames()</a></code>. Useful inside <code><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply()</a></code> which returns names but only when provided.</li>
</ul>
</div>
<div class="section level2">
<h2 id="dataframematrixtibble">data.frame/matrix/tibble<a class="anchor" aria-label="anchor" href="#dataframematrixtibble"></a>
</h2>
<ul>
<li>
<code><a href="reference/mixedSortDF.html">mixedSortDF()</a></code> - <code><a href="reference/mixedSort.html">mixedSort()</a></code> by columns or rownames</li>
<li>
<code><a href="reference/pasteByRow.html">pasteByRow()</a></code> - fast row-paste with delimiters, default skips blanks</li>
<li>
<code><a href="reference/pasteByRowOrdered.html">pasteByRowOrdered()</a></code> - nifty alternative that honors factor levels</li>
<li>
<code><a href="reference/rowGroupMeans.html">rowGroupMeans()</a></code>, <code><a href="reference/rowRmMadOutliers.html">rowRmMadOutliers()</a></code> - grouped row functions</li>
<li>
<code><a href="reference/mergeAllXY.html">mergeAllXY()</a></code> - merge a list of <code>data.frame</code> into one, keeping all rows</li>
<li>
<code><a href="reference/renameColumn.html">renameColumn()</a></code> - rename columns <code>from</code> and <code>to</code>.</li>
<li>
<code><a href="reference/kable_coloring.html">kable_coloring()</a></code> - flexible colorized <code>data.frame</code> output in Rmarkdown.</li>
</ul>
</div>
<div class="section level2">
<h2 id="string--grep">String / grep<a class="anchor" aria-label="anchor" href="#string--grep"></a>
</h2>
<ul>
<li>
<code><a href="reference/tcount.html">tcount()</a></code> - <code><a href="https://rdrr.io/r/base/table.html" class="external-link">table()</a></code> sorted high-to-low, with minimum count filter</li>
<li>
<code><a href="reference/middle.html">middle()</a></code> - show <code>n</code> entries from start, middle, then end.</li>
<li>
<code><a href="reference/gsubOrdered.html">gsubOrdered()</a></code> - <code><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub()</a></code> that returns ordered factor, inherits existing</li>
<li>
<code><a href="reference/gsubs.html">gsubs()</a></code> - <code><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub()</a></code> a vector of patterns/replacements.</li>
<li>
<code><a href="reference/grepls.html">grepls()</a></code> - grep the environment object names, including attached packages</li>
<li>
<code><a href="reference/vgrep.html">vgrep()</a></code>, <code><a href="reference/vigrep.html">vigrep()</a></code> - value-grep shortcut</li>
<li>
<code>unvgrep()</code>, <code><a href="reference/unvigrep.html">unvigrep()</a></code> - un-grep, remove matched results</li>
<li>
<code><a href="reference/provigrep.html">provigrep()</a></code> - progressive grep, returns matches in order of patterns</li>
<li>
<code><a href="reference/igrepHas.html">igrepHas()</a></code> - case-insensitive grep-any</li>
<li>
<code><a href="reference/ucfirst.html">ucfirst()</a></code> - upper-case the first letter of each word.</li>
<li>
<code><a href="reference/padString.html">padString()</a></code>, <code><a href="reference/padInteger.html">padInteger()</a></code> - produce strings from numeric values with consistent leading zeros.</li>
</ul>
</div>
<div class="section level2">
<h2 id="numeric">Numeric<a class="anchor" aria-label="anchor" href="#numeric"></a>
</h2>
<ul>
<li>
<code><a href="reference/formatInt.html">formatInt()</a></code> - opinionated <code><a href="https://rdrr.io/r/base/format.html" class="external-link">format()</a></code> for integers.</li>
<li>
<code><a href="reference/normScale.html">normScale()</a></code> - scale between 0 and 1 or custom range</li>
<li>
<code><a href="reference/noiseFloor.html">noiseFloor()</a></code> - apply noise floor, ceiling, with flexible replacements</li>
<li>
<code><a href="reference/log2signed.html">log2signed()</a></code>, <code><a href="reference/exp2signed.html">exp2signed()</a></code> - log2 with offset, and reciprocal</li>
<li>
<code><a href="reference/rowGroupMeans.html">rowGroupMeans()</a></code>, <code><a href="reference/rowRmMadOutliers.html">rowRmMadOutliers()</a></code> - efficient grouped row functions</li>
<li>
<code><a href="reference/deg2rad.html">deg2rad()</a></code>, <code><a href="reference/rad2deg.html">rad2deg()</a></code> - interconvert degrees and radians</li>
<li>
<code><a href="reference/rmNA.html">rmNA()</a></code> - remove NA values, with optional replacement</li>
<li>
<code><a href="reference/warpAroundZero.html">warpAroundZero()</a></code> - warp a numeric vector symmetrically around zero</li>
<li>
<code><a href="reference/rmInfinite.html">rmInfinite()</a></code> - remove infinite values, with optional replacement.</li>
<li>
<code><a href="reference/formatInt.html">formatInt()</a></code> - convenient <code><a href="https://rdrr.io/r/base/format.html" class="external-link">format()</a></code> for integer output, with comma-delimiter by default</li>
</ul>
<div class="section level3">
<h3 id="common-usage">Common usage<a class="anchor" aria-label="anchor" href="#common-usage"></a>
</h3>
<ul>
<li>convert zero to NA:</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/noiseFloor.html">noiseFloor</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">10</span>, minimum<span class="op">=</span><span class="fl">1e-20</span>, newValue<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] NA  1  2  3  4  5  6  7  8  9 10</span></span></code></pre></div>
<ul>
<li>convert values below floor to floor:</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/noiseFloor.html">noiseFloor</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">10</span>, minimum<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1]  3  3  3  3  4  5  6  7  8  9 10</span></span></code></pre></div>
<ul>
<li>convert values below floor or NA to floor:</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/noiseFloor.html">noiseFloor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">10</span>, <span class="cn">NA</span><span class="op">)</span>, minimum<span class="op">=</span><span class="fl">3</span>, adjustNA<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1]  3  3  3  3  4  5  6  7  8  9 10  3</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="practical--helpful">Practical / helpful<a class="anchor" aria-label="anchor" href="#practical--helpful"></a>
</h2>
<ul>
<li>
<code><a href="reference/jargs.html">jargs()</a></code> - pretty function arguments, optional pattern search argument name</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/jargs.html">jargs</a></span><span class="op">(</span><span class="va">plotSmoothScatter</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 x = ,</span></span>
<span><span class="co">#&gt;                 y = NULL,</span></span>
<span><span class="co">#&gt;              bwpi = 50,</span></span>
<span><span class="co">#&gt;             binpi = 50,</span></span>
<span><span class="co">#&gt;        bandwidthN = NULL,</span></span>
<span><span class="co">#&gt;              nbin = NULL,</span></span>
<span><span class="co">#&gt;            expand = c(0.04, 0.04),</span></span>
<span><span class="co">#&gt;       transFactor = 0.25,</span></span>
<span><span class="co">#&gt;    transformation = function( x ) x^transFactor,</span></span>
<span><span class="co">#&gt;              xlim = NULL,</span></span>
<span><span class="co">#&gt;              ylim = NULL,</span></span>
<span><span class="co">#&gt;              xlab = NULL,</span></span>
<span><span class="co">#&gt;              ylab = NULL,</span></span>
<span><span class="co">#&gt;          nrpoints = 0,</span></span>
<span><span class="co">#&gt;           colramp = c("white", "lightblue", "blue", "orange", "orangered2"),</span></span>
<span><span class="co">#&gt;               col = "black",</span></span>
<span><span class="co">#&gt;            doTest = FALSE,</span></span>
<span><span class="co">#&gt;    fillBackground = TRUE,</span></span>
<span><span class="co">#&gt;          naAction = c("remove", "floor0", "floor1"),</span></span>
<span><span class="co">#&gt;              xaxt = "s",</span></span>
<span><span class="co">#&gt;              yaxt = "s",</span></span>
<span><span class="co">#&gt;               add = FALSE,</span></span>
<span><span class="co">#&gt;               asp = NULL,</span></span>
<span><span class="co">#&gt; applyRangeCeiling = TRUE,</span></span>
<span><span class="co">#&gt;         useRaster = TRUE,</span></span>
<span><span class="co">#&gt;           verbose = FALSE,</span></span>
<span><span class="co">#&gt;               ... =</span></span></code></pre></div>
<ul>
<li>
<code><a href="reference/sdim.html">sdim()</a></code>, <code><a href="reference/sdim.html">ssdim()</a></code> - dimensions of list objects, or nested list of lists</li>
<li>
<code><a href="reference/sdim.html">sdima()</a></code> - runs <code><a href="reference/sdim.html">sdim()</a></code> on the attributes of an object.</li>
<li>
<code><a href="reference/isTRUEV.html">isTRUEV()</a></code>, <code><a href="reference/isFALSEV.html">isFALSEV()</a></code> - vectorized test for TRUE or FALSE values, since <code><a href="https://rdrr.io/r/base/Logic.html" class="external-link">isTRUE()</a></code> only operates on single values, and does not allow <code>NA</code>.</li>
<li>
<code><a href="reference/reload_rmarkdown_cache.html">reload_rmarkdown_cache()</a></code> - load RMarkdown cache folder into environment</li>
<li>
<code><a href="reference/call_fn_ellipsis.html">call_fn_ellipsis()</a></code> - for developers, call child function while passing only acceptable arguments in <code>...</code>. Instead of: <code>something(x, ...)</code>, use: <code>call_fn_ellipsis(something, x, ...)</code> and never worry about <code>...</code>.</li>
<li>
<code><a href="reference/log2signed.html">log2signed()</a></code>, <code><a href="reference/exp2signed.html">exp2signed()</a></code> - convenient <code>log2(1 + x)</code> or its reciprocal, using customizable offset.</li>
<li>
<code><a href="reference/newestFile.html">newestFile()</a></code> - most recently modified file from a vector of files</li>
</ul>
<div class="section level3">
<h3 id="r-console">R console<a class="anchor" aria-label="anchor" href="#r-console"></a>
</h3>
<ul>
<li>
<code><a href="reference/jargs.html">jargs()</a></code> - Jam argument list - see “Practical” above for example</li>
<li>
<code><a href="reference/lldf.html">lldf()</a></code> - <code><a href="https://rdrr.io/r/base/ls.html" class="external-link">ls()</a></code> with <code><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size()</a></code> into <code>data.frame</code>
</li>
<li>
<code><a href="reference/middle.html">middle()</a></code> - Similar to <code><a href="https://rdrr.io/r/utils/head.html" class="external-link">head()</a></code> and <code><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail()</a></code>, <code><a href="reference/middle.html">middle()</a></code> shows <code>n</code> entries from beginning, middle, to end.</li>
<li>
<code><a href="reference/printDebug.html">printDebug()</a></code> - colorized text output</li>
<li>
<code><a href="reference/setPrompt.html">setPrompt()</a></code> - colorized R console prompt with project name and R version</li>
</ul>
</div>
<div class="section level3">
<h3 id="rmarkdown">RMarkdown<a class="anchor" aria-label="anchor" href="#rmarkdown"></a>
</h3>
<ul>
<li><p><code><a href="reference/reload_rmarkdown_cache.html">reload_rmarkdown_cache()</a></code> - when rendering RMarkdown with <code>cache=TRUE</code>, this function reads the cache to reload the environment without re-processing, to recover the exact result for continued work.</p></li>
<li>
<p><code><a href="reference/printDebug.html">printDebugHtml()</a></code> - colored HTML output.</p>
<ul>
<li>Shortcut for <code>printDebug(..., htmlOut=TRUE, comments=FALSE)</code>, or <code>options("jam.htmlOut"=TRUE, "jam.comment"=FALSE)</code>.</li>
<li>The RMarkdown chunk must include: <code>results='asis'</code>
</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/printDebug.html">printDebugHtml</a></span><span class="op">(</span><span class="st">"printDebugHtml(): "</span>,</span>
<span>  <span class="st">"Output is colorized: "</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">LETTERS</span>, <span class="fl">8</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>(<span style="color:#00C3C3FF">12:05:41</span>) 07Mar2025: <span style="color:#FF7F00FF">printDebugHtml(): </span><span style="color:#1E90FFFF">Output is colorized: </span><span style="color:#FF7F00FF">A,</span><span style="color:#ED983BFF">B,</span><span style="color:#FF7F00FF">C,</span><span style="color:#ED983BFF">D,</span><span style="color:#FF7F00FF">E,</span><span style="color:#ED983BFF">F,</span><span style="color:#FF7F00FF">G,</span><span style="color:#ED983BFF">H</span><br></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">withr</span><span class="fu">::</span><span class="fu"><a href="https://withr.r-lib.org/reference/with_options.html" class="external-link">with_options</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>jam.htmlOut<span class="op">=</span><span class="cn">TRUE</span>, jam.comment<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>, <span class="op">{</span></span>
<span>  <span class="fu"><a href="reference/printDebug.html">printDebugHtml</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"printDebug() using withr::with_options(): "</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Output should be colorized: "</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">LETTERS</span>, <span class="fl">8</span><span class="op">)</span><span class="op">)</span>;</span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>(<span style="color:#00C3C3FF">12:05:41</span>) 07Mar2025: <span style="color:#FF7F00FF">printDebug() using withr::with_options(): </span><span style="color:#1E90FFFF">Output should be colorized: </span><span style="color:#FF7F00FF">A,</span><span style="color:#ED983BFF">B,</span><span style="color:#FF7F00FF">C,</span><span style="color:#ED983BFF">D,</span><span style="color:#FF7F00FF">E,</span><span style="color:#ED983BFF">F,</span><span style="color:#FF7F00FF">G,</span><span style="color:#ED983BFF">H</span><br></p>
<ul>
<li>
<p><code><a href="reference/kable_coloring.html">kable_coloring()</a></code> - applies categorical colors to <code>kable()</code> output using <code><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kableExtra::kable()</a></code>.</p>
<ul>
<li>It also applies a contrasting text color.</li>
<li>Unfortunately, the HTML output is not compatible with this page on Github, see package function docs in RStudio.</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expt_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  Sample_ID<span class="op">=</span><span class="st">""</span>,</span>
<span>  Treatment<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Vehicle"</span>, <span class="st">"Dex"</span><span class="op">)</span>, each<span class="op">=</span><span class="fl">6</span><span class="op">)</span>,</span>
<span>  Genotype<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Wildtype"</span>, <span class="st">"Knockout"</span><span class="op">)</span>, each<span class="op">=</span><span class="fl">3</span><span class="op">)</span>,</span>
<span>  Rep<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"rep"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">expt_df</span><span class="op">$</span><span class="va">Sample_ID</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/pasteByRow.html">pasteByRow</a></span><span class="op">(</span><span class="va">expt_df</span><span class="op">[</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># define colors</span></span>
<span><span class="va">colorSub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Vehicle<span class="op">=</span><span class="st">"palegoldenrod"</span>,</span>
<span>  Dex<span class="op">=</span><span class="st">"navy"</span>,</span>
<span>  Wildtype<span class="op">=</span><span class="st">"gold"</span>,</span>
<span>  Knockout<span class="op">=</span><span class="st">"firebrick"</span>,</span>
<span>  <span class="fu"><a href="reference/nameVector.html">nameVector</a></span><span class="op">(</span><span class="fu"><a href="reference/color2gradient.html">color2gradient</a></span><span class="op">(</span><span class="st">"grey48"</span>, n<span class="op">=</span><span class="fl">3</span>, dex<span class="op">=</span><span class="fl">10</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"rep"</span>, <span class="fl">3</span><span class="op">)</span>, suffix<span class="op">=</span><span class="st">""</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="reference/nameVector.html">nameVector</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="reference/color2gradient.html">color2gradient</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">3</span>,</span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"goldenrod1"</span>, <span class="st">"indianred3"</span>, <span class="st">"royalblue3"</span>, <span class="st">"darkorchid4"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="va">expt_df</span><span class="op">$</span><span class="va">Sample_ID</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">kbl</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/kable_coloring.html">kable_coloring</a></span><span class="op">(</span></span>
<span>  <span class="va">expt_df</span>,</span>
<span>  caption<span class="op">=</span><span class="st">"Experiment design table showing categorical color assignment."</span>,</span>
<span>  <span class="va">colorSub</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="other-related-jam-packages">Other related Jam packages<a class="anchor" aria-label="anchor" href="#other-related-jam-packages"></a>
</h2>
<p>Jam Github R packages are being transitioned to CRAN/Bioconductor:</p>
<ul>
<li>
<code>venndir</code>: Venn diagrams with direction, designed for published figures.</li>
<li>
<code>multienrichjam</code>: Multi-enrichment pathway analysis and visualization tools.</li>
<li>
<code>splicejam</code>: Sashimi plots for RNA-seq coverage and junction data.</li>
<li>
<code>jamma</code>: MA-plots as a unified <strong><em>data signal</em></strong> quality control toolset.</li>
<li>
<code>colorjam</code>: <code>rainbowJam()</code>, Categorical colors with improved visual contrast.</li>
<li>
<code>genejam</code>: Fast, structured approach to gene symbol integration.</li>
<li>
<code>platjam</code>: Platform specific functions: Nanostring, Salmon, Proteomics, Lipidomics; NGS coverage heatmaps.</li>
<li>
<code>jamses</code>: <code>heatmap_se()</code> friendly wrapper for ComplexHeatmap; other integrated methods for factor-aware design/contrasts, normalization, contrasts, heatmaps.</li>
<li>
<code>jamsession</code>: properly save/load R objects, R sessions, R functions.</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/jmw86069/jamba/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/jmw86069/jamba/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing jamba</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>James M. Ward <br><small class="roles"> Author, maintainer, copyright holder </small> <a href="https://orcid.org/0000-0002-9510-2848" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by James M. Ward.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
